refactor(compiler): introduce block parsing in lexer

crisbeto opened the pull request:
⚠️Disclaimer⚠️ this PR implements syntax that is still in an open RFC. It will be adjusted once the RFC is closed.

These changes extend the lexer to recognize the concepts of a block group (`{#foo paramA; paramB}{/foo}`) and a block (`{:foo paramA; paramB;}`) which will be useful later on for the control flow and defer proposals. Block groups can be used anywhere and require a closing tag while block can only be used inside of a block group.

The idea is that in the next PRs the markup AST will be expanded to have some more specialized nodes like `ConditionalBlock` or `DeferBlock` which will then be turned into instructions.

AndrewKushnir commented on packages/compiler/test/ml_parser/lexer_spec.ts:
```diff
@@ -1814,6 +1814,221 @@ import {ParseLocation, ParseSourceFile, ParseSourceSpan} from '../../src/parse_u
+        expect(tokenizeBlock(
+                   '{#if expr}' +
+                   'foo' +
+                   '{:else if otherExpr === 1}' +
```
That looks reasonable. We'd probably need to discuss if we want to have `{:elseif ...}` instead, so we have more complete picture after tokenizing an input.

AndrewKushnir reviewed (approved):
@crisbeto looks great, thanks 👍

AndrewKushnir commented on packages/compiler/src/ml_parser/tokens.ts:
```diff
@@ -33,15 +33,22 @@ export const enum TokenType {
   EXPANSION_FORM_END,
-  EOF
+  EOF,
+  BLOCK_OPEN_START,
```
"Block" looks good to me, we can also add JSDoc here (for each new token) to explain what it represents (so that there is no confusion).

crisbeto commented:
After chatting with @alxhub, I've updated the terminology so `{#foo}{/foo}` represents a block group while `{:foo}` is a block. Everything else is the same.

crisbeto commented:
Caretaker note: the presubmit passes, except for a couple of already-failing targets.

crisbeto committed:
refactor(compiler): introduce block parsing in lexer

⚠️Disclaimer⚠️ this PR implements syntax that is still in an open RFC. It will be adjusted once the RFC is closed.

These changes extend the lexer to recognize the concepts of a block group (`{#foo paramA; paramB}{/foo}`) and a block (`{:foo paramA; paramB;}`) which will be useful later on for the control flow and defer proposals. Block groups can be used anywhere and require a closing tag while block can only be used inside of a block.

The idea is that in the next PRs the markup AST will be expanded to have some more specialized node like `ConditionalBlock` or `DeferBlock` which will then be turned into instructions.

crisbeto commented:
I've pushed one more tweak, because I realized that the source map locations were slightly off.

crisbeto commented:
Caretaker note: the presubmit failures are due to an unrelated issue in one of the targets from the blueprint.
